go/libraries/doltcore/sqle/dfunctions/dolt_pull.go

```diff
@@ -0,0 +1,213 @@
+}
+
+func (d DoltPullFunc) Type() sql.Type {
+	return sql.Boolean
+}
+
+func (d DoltPullFunc) WithChildren(ctx *sql.Context, children ...sql.Expression) (sql.Expression, error) {
+	return NewPullFunc(ctx, children...)
+}
+
+func (d DoltPullFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	dbName := ctx.GetCurrentDatabase()
+
+	if len(dbName) == 0 {
+		return noConflicts, fmt.Errorf("empty database name.")
+	}
+
+	sess := dsess.DSessFromSess(ctx.Session)
+	dbData, ok := sess.GetDbData(ctx, dbName)
+	if !ok {
+		return noConflicts, sql.ErrDatabaseNotFound.New(dbName)
+	}
+
+	ap := cli.CreatePullArgParser()
+	args, err := getDoltArgs(ctx, row, d.Children())
+
+	apr, err := ap.Parse(args)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	if apr.NArg() > 1 {
+		return noConflicts, actions.ErrInvalidPullArgs
+	}
+
+	var remoteName string
+	if apr.NArg() == 1 {
+		remoteName = apr.Arg(0)
+	}
+
+	pullSpec, err := env.NewPullSpec(ctx, dbData.Rsr, remoteName, apr.Contains(cli.SquashParam), apr.Contains(cli.NoFFParam), apr.Contains(cli.ForceFlag))
+	if err != nil {
+		return noConflicts, err
+	}
+
+	srcDB, err := pullSpec.Remote.GetRemoteDBWithoutCaching(ctx, dbData.Ddb.ValueReadWriter().Format())
+	if err != nil {
+		return noConflicts, fmt.Errorf("failed to get remote db; %w", err)
+	}
+
+	ws, err := sess.WorkingSet(ctx, dbName)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	var conflicts interface{}
+	for _, refSpec := range pullSpec.RefSpecs {
+		remoteTrackRef := refSpec.DestRef(pullSpec.Branch)
+
+		if remoteTrackRef != nil {
+
+			// todo: can we pass nil for either of the channels?
+			srcDBCommit, err := actions.FetchRemoteBranch(ctx, dbData.Rsw.TempTableFilesDir(), pullSpec.Remote, srcDB, dbData.Ddb, pullSpec.Branch, remoteTrackRef, runProgFuncs, stopProgFuncs)
+			if err != nil {
+				return noConflicts, err
+			}
+
+			// TODO: this could be replaced with a canFF check to test for error
+			err = dbData.Ddb.FastForward(ctx, remoteTrackRef, srcDBCommit)
+			if err != nil {
+				return noConflicts, fmt.Errorf("fetch failed; %w", err)
+			}
+
+			roots, ok := sess.GetRoots(ctx, dbName)
+			if !ok {
+				return noConflicts, sql.ErrDatabaseNotFound.New(dbName)
+			}
+
+			mergeSpec, err := createMergeSpec(ctx, sess, dbName, apr, remoteTrackRef.String())
+			if err != nil {
+				return noConflicts, err
+			}
+			ws, conflicts, err = mergeIntoWorkingSet(ctx, sess, roots, ws, dbName, mergeSpec)
+			if err != nil && !errors.Is(doltdb.ErrUpToDate, err) {
+				return conflicts, err
+			}
+
+			err = sess.SetWorkingSet(ctx, dbName, ws, nil)
+			if err != nil {
+				return conflicts, err
+			}
+		}
+	}
+
+	err = actions.FetchFollowTags(ctx, dbData.Rsw.TempTableFilesDir(), srcDB, dbData.Ddb, runProgFuncs, stopProgFuncs)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	return noConflicts, nil
+}
+
+func pullerProgFunc(ctx context.Context, pullerEventCh <-chan datas.PullerEvent) {
+	for {
```

Comment:
I don't understand the point of having two selects in this loop like this. What am I missing?

Reply:
i've only seen short circuits implemented this way in the docs i've read. granted i don't understand the tradeoffs of short circuiting (the extra select) vs a 50/50 chance of exiting if we can both read from the channel and exit (second select)